kong/kong.lua

```diff
@@ -252,6 +252,18 @@ function Kong.balancer()
   end
 end
 
+function Kong.rewrite()
+  core.rewrite.before()
+
+  for _, plugin in ipairs(singletons.loaded_plugins) do
+    if plugin.schema.no_consumer and plugin.schema.no_api then
+      plugin.handler:rewrite()
+    end
+  end
+
+  core.rewrite.after()
+end
+
```

Comment:
> the above runs on loaded_plugins which means that it will run all plugins available, no matter what.

I understand your point but it would be over-optimization IMHO. Currently every other handler also runs on every plugin regardless if they have an event handler associated or not (ie, the `access()` for every plugin is invoked regardless if the plugin implements that method).

> btw, can't we bring the api detection forward, to run in the core.rewrite.before() handler? not sure what it takes, but then we can run plugins like in any phase, including "per api" configured plugins

My use-case is to implement URL rewriting **before** the APIs are matched, and rewrite the URI to be matched later by the router.